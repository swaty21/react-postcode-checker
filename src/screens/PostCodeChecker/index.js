import React, { useState } from 'react';
import Button from '../../components/Button';
import Input from '../../components/Input';
import { VALID_LSOA, VALID_POST_CODE } from '../../config/postCodeCheckConfig';
import { fetchPostCodeDetails } from '../../api/postCodeApi';

const PostCodeChecker = () => {
  const [postCodeValue, setInputPostCode] = useState('');
  const [showPostCodeMessage, setShowPostCodeMessage] = useState(false);
  const [isPostCodeValid, setPostCodeValidation] = useState(false);

  const handleInputChange = (event) => {
    setInputPostCode(event.target.value);
  };

  const handlePostCodeValidation = (response) => {
    const postCode = VALID_LSOA.find((lsoa) => {
      return response.result.lsoa.indexOf(lsoa) > -1;
    });
    if (postCode === undefined) {
      return setPostCodeValidation(false);
    }
    return setPostCodeValidation(true);
  };

  const handleSubmitClick = async () => {
    const postCodeWithoutSpace = postCodeValue.replace(' ', '');

    const postCodeCheck = VALID_POST_CODE.find((postcode) => {
      const abc = postcode.replace(' ', '');
      return postCodeWithoutSpace.indexOf(abc) > -1;
    });
    if (postCodeCheck) {
      setShowPostCodeMessage(true);
      setPostCodeValidation(true);
      return;
    }
    try {
      setShowPostCodeMessage(true);
      const response = await fetchPostCodeDetails(postCodeWithoutSpace);
      handlePostCodeValidation(response.data);
    } catch (error) {
      setShowPostCodeMessage(true);
    }
  };

  return (
    <>
      <h3> Please Enter Your Postcode</h3>
      <Input id="input-postcode" onChange={handleInputChange} value={postCodeValue} />
      <Button id="submit" onClick={handleSubmitClick} name="Submit" type="Submit" />
      {showPostCodeMessage &&
        (isPostCodeValid ? (
          <p data-testid="success-postcode-msg">We support the postcode.</p>
        ) : (
          <p data-testid="notfound-postcode-msg">Sorry We dont support the postcode.</p>
        ))}
    </>
  );
};

export default PostCodeChecker;
