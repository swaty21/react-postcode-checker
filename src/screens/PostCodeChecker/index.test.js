/**
 * @jest-environment jsdom
 */
import React from 'react';
import axios from 'axios';
import { render, fireEvent, waitFor, act } from '@testing-library/react';
import PostCodeChecker from './index';

jest.mock('axios');

describe('PostCodeChecker', () => {
  it('should render PostCodeChecker component', () => {
    const { container } = render(<PostCodeChecker />);
    expect(container).toMatchSnapshot();
  });
  it('should show success msg when valid postCode', () => {
    const { container, getByTestId } = render(<PostCodeChecker />);
    const inputElement = getByTestId('input-postcode');
    const btnElement = getByTestId('submit');
    fireEvent.change(inputElement, { target: { value: 'SE1 7QD' } });
    fireEvent.click(btnElement);
    expect(container).toMatchSnapshot();
  });
  it('should call the API to show postcode is invalid', () => {
    const { container, getByTestId } = render(<PostCodeChecker />);
    const postCodeDetails = {
      data: {
        result: {
          postcode: 'G41 1AL',
          lsoa: 'Pollokshields East - 02',
          msoa: 'Pollokshields East'
        }
      }
    };
    axios.get.mockResolvedValueOnce(postCodeDetails);
    const inputElement = getByTestId('input-postcode');
    const btnElement = getByTestId('submit');
    fireEvent.change(inputElement, { target: { value: 'G41 1AL' } });
    fireEvent.click(btnElement);
    expect(container).toMatchSnapshot();
  });
  it('should show success msg when valid  lsoa', async () => {
    const { container, getByTestId } = render(<PostCodeChecker />);
    const postCodeDetails = {
      data: {
        result: {
          postcode: 'G41 1AM',
          lsoa: 'Southwark East - 02',
          msoa: 'Pollokshields East'
        }
      }
    };
    axios.get.mockResolvedValueOnce(postCodeDetails);
    const inputElement = getByTestId('input-postcode');
    const btnElement = getByTestId('submit');
    act(() => {
      fireEvent.change(inputElement, { target: { value: 'G41 1AM' } });
    });
    await waitFor(() => {
      act(() => {
        fireEvent.click(btnElement);
      });
    });
    expect(container).toMatchSnapshot();
  });
});
