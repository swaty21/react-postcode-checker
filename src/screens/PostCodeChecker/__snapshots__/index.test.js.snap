// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCodeChecker should call the API to show postcode is invalid 1`] = `
<div>
  <h3>
     Please Enter Your Postcode
  </h3>
  <input
    data-testid="input-postcode"
    value="G41 1AL"
  />
  <button
    data-testid="submit"
    name="Submit"
    type="Submit"
  >
    Submit
     
  </button>
  <p
    data-testid="notfound-postcode-msg"
  >
    Sorry We dont support the postcode.
  </p>
</div>
`;

exports[`PostCodeChecker should render PostCodeChecker component 1`] = `
<div>
  <h3>
     Please Enter Your Postcode
  </h3>
  <input
    data-testid="input-postcode"
    value=""
  />
  <button
    data-testid="submit"
    name="Submit"
    type="Submit"
  >
    Submit
     
  </button>
</div>
`;

exports[`PostCodeChecker should show success msg when valid  lsoa 1`] = `
<div>
  <h3>
     Please Enter Your Postcode
  </h3>
  <input
    data-testid="input-postcode"
    value="G41 1AM"
  />
  <button
    data-testid="submit"
    name="Submit"
    type="Submit"
  >
    Submit
     
  </button>
  <p
    data-testid="success-postcode-msg"
  >
    We support the postcode.
  </p>
</div>
`;

exports[`PostCodeChecker should show success msg when valid postCode 1`] = `
<div>
  <h3>
     Please Enter Your Postcode
  </h3>
  <input
    data-testid="input-postcode"
    value="SE1 7QD"
  />
  <button
    data-testid="submit"
    name="Submit"
    type="Submit"
  >
    Submit
     
  </button>
  <p
    data-testid="success-postcode-msg"
  >
    We support the postcode.
  </p>
</div>
`;
